design/variables.md

```diff
@@ -0,0 +1,71 @@
+# Variables context and condition simplification
+
+## Processing within a file
+
+Changes:
+- Add variables to the context, as they are read
+- Add `else` and `elif`
+
+Example jobs template:
+
+```yaml
+parameters:
+  jobName: ''
+jobs:
+- job: ${{ parameters.jobName }}
+  variables:
+    public: ${{ eq(variables['System.TeamProject'], 'public') }}
+    publicOrPR: ${{ or(eq(variables.public, 'true'), eq(variables['Build.Reason'], 'PullRequest')) }}
```

Comment:
I imagine that switching between access methods will cause confusion, ie, people will want to either always do a hash table lookup method (`variables['key']`) or an object property method (`variables.key`).  Asking people to understand the context and switch between the methods may lead to frustration.

Reply:
they are interchangeable, but the dot notation is restricted (member name must consist of a-z0-9_)

Reply:
...and that is current behavior already. Nothing is changing there.

Reply:
Eric, I actually ran into a bug yesterday which proved that they are not interchangeable. Only the hash table syntax works every time, the dot notation does not. Surprised me :)

Reply:
Do you have a sample of the bug?  My understanding (based on ad-hoc testing) is that they are interchangeable modulo the caveat which Eric outlined above.

Reply:
Yep, use any of the template expressions in your yaml:

${{ if and(ne(variables['System.TeamProject'], 'public'), notin(variables['Build.Reason'], 'PullRequest'), not(contains(variables['Build.DefinitionName'], 'github'))) }}:

Replace all of these with the dot notation and it doesn't work as it should. Switch it back to hash table notation and it does.

Reply:
At least that was my experience. I much prefer the dot notation because it's less clutter, less typing, and more in line with json and normal property lookups. So I tried switching them but it didn't work properly.

Reply:
> member name must consist of a-z0-9_

`.` is treated differently in "dot" notation.